plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSnapshotPageSource.java

```diff
@@ -0,0 +1,224 @@
+                columnHandles.stream().map(HiveColumnHandle::getHiveType).toList(), false);
+        this.pageBuilder = new PageBuilder(dataHandles.stream().map(HiveColumnHandle::getType).toList());
+        Map<String, String> nameToPartitionValueMap = partitionKeyList.stream().collect(
+                Collectors.toMap(e -> e.name(), e -> e.value()));
+        this.partitionValueMap = new HashMap<>();
+        for (int i = 0; i < dataHandles.size(); i++) {
+            HiveColumnHandle handle = dataHandles.get(i);
+            if (handle.isPartitionKey()) {
+                partitionValueMap.put(i + HOODIE_META_COLUMNS.size(), nameToPartitionValueMap.get(handle.getName()));
+            }
+        }
+        this.avroSerializer = new HudiAvroSerializer(columnHandles);
+        this.recordKeyFieldPos = 2;
+        preCombineField.ifPresent(s -> this.payloadProps.setProperty(HoodiePayloadProps.PAYLOAD_ORDERING_FIELD_PROP_KEY, s));
+    }
+
+    @Override
+    public long getCompletedBytes() {
+        return 0;
+    }
+
+    @Override
+    public long getReadTimeNanos() {
+        return 0;
+    }
+
+    @Override
+    public boolean isFinished() {
+        return (baseFilePageSource.isEmpty() || baseFilePageSource.get().isFinished())
+                && (logRecordMap != null && logRecordMap.isEmpty());
+    }
+
+    @Override
+    public CompletableFuture<?> isBlocked()
+    {
+        if (baseFilePageSource.isPresent()) {
+            return baseFilePageSource.get().isBlocked();
+        }
+        return CompletableFuture.completedFuture(0);
+    }
+
+
+    @Override
+    public OptionalLong getCompletedPositions()
+    {
+        if (baseFilePageSource.isPresent()) {
+            return baseFilePageSource.get().getCompletedPositions();
+        }
+        return OptionalLong.of(10);
+    }
+
+    @Override
+    public Metrics getMetrics()
+    {
+        if (baseFilePageSource.isPresent()) {
+            return baseFilePageSource.get().getMetrics();
+        }
+        return Metrics.EMPTY;
+    }
+
+    @Override
+    public Page getNextPage() {
+        if (logRecordMap == null) {
+            try (HoodieMergedLogRecordScanner logScanner = getMergedLogRecordScanner(storage, basePath, split, readerSchema)) {
+                logRecordMap = logScanner.getRecords();
+            } catch (IOException e) {
+                throw new HoodieIOException("Cannot read Hudi split " + split, e);
+            }
+        }
+
+        checkState(pageBuilder.isEmpty(), "PageBuilder is not empty at the beginning of a new page");
+
+        int size = columnHandles.size();
+        if (baseFilePageSource.isPresent()) {
+            Page page = baseFilePageSource.get().getNextPage();
+            if (page != null) {
+                try {
+                    // Merge records from the page with log records
+                    for (int pos = 0; pos < page.getPositionCount(); pos++) {
```

Comment:
Is there a way to do batch and parallel merge instead of doing sequentially per page?

Reply:
The current logic follows the same log merging procedure, which is sequential for log merging.  Since multiple splits are parallelized across task, this is OK for now.